Source/Csla.Blazor.WebAssembly/ApplicationContextManager.cs

```diff
@@ -11,180 +11,177 @@
+  protected AuthenticationStateProvider AuthenticationStateProvider { get; }
 
-    /// <summary>
-    /// Gets or sets a reference to the current ApplicationContext.
-    /// </summary>
-    public ApplicationContext ApplicationContext { get; set; }
+  /// <summary>
+  /// Gets or sets a reference to the current ApplicationContext.
+  /// </summary>
+  public ApplicationContext ApplicationContext { get; set; }
 
-    private void InitializeUser()
-    {
-      AuthenticationStateProvider_AuthenticationStateChanged(AuthenticationStateProvider.GetAuthenticationStateAsync());
-    }
+  private void InitializeUser()
+  {
+    AuthenticationStateProvider_AuthenticationStateChanged(AuthenticationStateProvider.GetAuthenticationStateAsync());
+  }
 
-    private void AuthenticationStateProvider_AuthenticationStateChanged(Task<AuthenticationState> task)
-    {
-      _currentPrincipal = null;
-      AuthenticationState = task;
-    }
+  private void AuthenticationStateProvider_AuthenticationStateChanged(Task<AuthenticationState> task)
+  {
+    AuthenticationState = task;
+  }
 
-    /// <summary>
-    /// Gets a value indicating whether this
-    /// context manager is valid for use in
-    /// the current environment.
-    /// </summary>
-    public bool IsValid => true;
-
-    /// <summary>
-    /// Gets a value indicating whether the current runtime
-    /// is stateful (e.g. WPF, Blazor, etc.)
-    /// </summary>
-    public bool IsStatefulContext => true;
-
-    /// <summary>
-    /// Gets the current principal.
-    /// </summary>
-    public IPrincipal GetUser()
-    {
-      if (_currentPrincipal == null)
-      {
-        if (AuthenticationState.IsCompleted && AuthenticationState.Result != null)
-          _currentPrincipal = AuthenticationState.Result.User;
-        else
-          _currentPrincipal = new ClaimsPrincipal();
-      }
-      return _currentPrincipal;
-    }
+  /// <summary>
+  /// Gets a value indicating whether this
+  /// context manager is valid for use in
+  /// the current environment.
+  /// </summary>
+  public bool IsValid => true;
 
-    /// <summary>
-    /// Sets the current principal ONLY IN APPLICATIONCONTEXT.
-    /// To set the value correctly, use your specific
-    /// ApplicationContextStateProvider implementation.
-    /// </summary>
-    /// <param name="principal">Principal object.</param>
-    public virtual void SetUser(IPrincipal principal)
-    {
-      throw new NotSupportedException(nameof(SetUser));
-    }
+  /// <summary>
+  /// Gets a value indicating whether the current runtime
+  /// is stateful (e.g. WPF, Blazor, etc.)
+  /// </summary>
+  public bool IsStatefulContext => true;
+
+  private static readonly IPrincipal _anonymous = new ClaimsPrincipal(new ClaimsIdentity());
+
+  /// <summary>
+  /// Gets the current principal.
+  /// </summary>
+  public IPrincipal GetUser()
+  {
+    IPrincipal result;
+    if (AuthenticationState.IsCompleted && AuthenticationState.Result != null)
```

Comment:
Shouldn't you test for `IsCompletedSuccessfully` instead of `IsCompleted`. The latter can also mean cancelled or faulted (exception). I'm not sure if this would result in an exception here.

Reply:
I think that's a good improvement. I believe the null check prevents an exception, but this will be better.